File: engines/r-scripts/descriptives.R
# descriptives.R - Descriptive statistics for Method Studio
#
# Expected environment variables (set by wrapper.R from request data):
#   variables : character vector of column names to analyze
#   + columnar data with those column names present
#
# Returns a named list with stats for each variable

if (!exists("variables") || length(variables) == 0) {
  stop("Variable 'variables' is required - specify column names to analyze")
}

# Variables come as a list from JSON, extract as character vector
var_names <- if (is.list(variables)) unlist(variables) else variables

# Helper functions
calc_mode <- function(v) {
  tbl <- table(v)
  modes <- as.numeric(names(tbl)[tbl == max(tbl)])
  if (length(modes) == 1) modes[1] else modes
}

describe_variable <- function(var_name) {
  if (!exists(var_name)) {
    stop(paste0("Column '", var_name, "' not found in data"))
  }

  # Get data - may come as list from JSON
  x_raw <- get(var_name)
  x_all <- as.numeric(if (is.list(x_raw)) unlist(x_raw) else x_raw)
  x_valid <- x_all[!is.na(x_all)]

  n <- length(x_all)
  n_missing <- sum(is.na(x_all))
  n_valid <- length(x_valid)

  if (n_valid == 0) {
    return(list(
      n = n,
      missing = n_missing,
      mean = NA,
      median = NA,
      std = NA,
      variance = NA,
      se_mean = NA,
      min = NA,
      max = NA,
      range = NA,
      q1 = NA,
      q3 = NA,
      iqr = NA,
      skewness = NA,
      kurtosis = NA
    ))
  }

  # Central tendency
  x_mean <- mean(x_valid)
  x_median <- median(x_valid)

  # Dispersion
  x_sd <- if (n_valid > 1) sd(x_valid) else 0
  x_var <- if (n_valid > 1) var(x_valid) else 0
  x_min <- min(x_valid)
  x_max <- max(x_valid)
  x_range <- x_max - x_min
  x_se <- if (n_valid > 0) x_sd / sqrt(n_valid) else 0

  # Quartiles
  quarts <- quantile(x_valid, probs = c(0.25, 0.75))
  x_q1 <- quarts[1]
  x_q3 <- quarts[2]
  x_iqr <- IQR(x_valid)

  # Shape
  if (requireNamespace("psych", quietly = TRUE) && n_valid >= 3) {
    x_skewness <- psych::skew(x_valid)
    x_kurtosis <- if (n_valid >= 4) psych::kurtosi(x_valid) else NA
  } else {
    # Fallback: manual calculation
    if (n_valid >= 3) {
      m3 <- mean((x_valid - x_mean)^3)
      x_skewness <- m3 / x_sd^3
    } else {
      x_skewness <- NA
    }
    if (n_valid >= 4) {
      m4 <- mean((x_valid - x_mean)^4)
      x_kurtosis <- (m4 / x_sd^4) - 3
    } else {
      x_kurtosis <- NA
    }
  }

  list(
    n = n_valid,
    missing = n_missing,
    mean = round(x_mean, 6),
    median = round(x_median, 6),
    std = round(x_sd, 6),
    variance = round(x_var, 6),
    se_mean = round(x_se, 6),
    min = round(x_min, 6),
    max = round(x_max, 6),
    range = round(x_range, 6),
    q1 = round(unname(x_q1), 6),
    q3 = round(unname(x_q3), 6),
    iqr = round(x_iqr, 6),
    skewness = if (!is.na(x_skewness)) round(x_skewness, 6) else NA,
    kurtosis = if (!is.na(x_kurtosis)) round(x_kurtosis, 6) else NA
  )
}

# Process each variable
stats_out <- list()
for (var_name in var_names) {
  stats_out[[var_name]] <- describe_variable(var_name)
}

# Create histogram for first variable
if (length(var_names) > 0) {
  first_var <- var_names[1]
  x_raw <- get(first_var)
  x_data <- as.numeric(if (is.list(x_raw)) unlist(x_raw) else x_raw)
  x_valid <- x_data[!is.na(x_data)]

  if (length(x_valid) > 0) {
    hist(x_valid,
         main = paste("Distribution of", first_var),
         xlab = first_var,
         col = "#4C72B0",
         border = "white",
         probability = TRUE)
    if (length(x_valid) > 1) {
      lines(density(x_valid), col = "#DD8452", lwd = 2)
    }
  }
}

# Compose result
result <- list(
  variables = var_names,
  stats = stats_out
)

result

File: engines/python-scripts/descriptives.py
"""
Descriptive Statistics Script
==============================
Expects the following variables in the execution namespace (injected by the engine):

  variables : list[str]
      Column names to analyze
  data : dict
      Columnar data dictionary with column names as keys and arrays as values

Produces a 'result' dict with the following structure:

  {
    "variables": ["x", "y", ...],
    "stats": {
      "x": {
        "n": int,
        "missing": int,
        "mean": float,
        "median": float,
        "std": float,
        "variance": float,
        "se_mean": float,
        "min": float,
        "max": float,
        "range": float,
        "q1": float,
        "q3": float,
        "iqr": float,
        "skewness": float,
        "kurtosis": float,
      },
      ...
    }
  }
"""

import numpy as np
from scipy import stats as scipy_stats


def _coerce(val) -> np.ndarray:
    """Convert JSON value to numpy array (keeping NaN for missing detection)."""
    if isinstance(val, np.ndarray):
        return val.astype(float)
    if isinstance(val, (list, tuple)):
        return np.array(val, dtype=float)
    return np.array([val], dtype=float)


def _get_column(col_name: str, data_dict: dict) -> np.ndarray:
    """Get column data from the data dictionary."""
    if col_name not in data_dict:
        raise ValueError(f"Column '{col_name}' not found in data")
    return _coerce(data_dict[col_name])


def _describe_column(arr: np.ndarray) -> dict:
    """Compute descriptive statistics for a numeric array."""
    n_total = len(arr)
    mask = ~np.isnan(arr)
    clean = arr[mask]
    n_valid = len(clean)
    n_missing = n_total - n_valid

    if n_valid == 0:
        return {
            "n": 0,
            "missing": int(n_missing),
            "mean": None,
            "median": None,
            "std": None,
            "variance": None,
            "se_mean": None,
            "min": None,
            "max": None,
            "range": None,
            "q1": None,
            "q3": None,
            "iqr": None,
            "skewness": None,
            "kurtosis": None,
        }

    mean = float(np.mean(clean))
    median = float(np.median(clean))
    std = float(np.std(clean, ddof=1)) if n_valid > 1 else 0.0
    variance = float(np.var(clean, ddof=1)) if n_valid > 1 else 0.0
    se_mean = std / np.sqrt(n_valid) if n_valid > 0 else 0.0
    vmin = float(np.min(clean))
    vmax = float(np.max(clean))

    q1, q3 = float(np.percentile(clean, 25)), float(np.percentile(clean, 75))
    iqr = q3 - q1

    skewness = float(scipy_stats.skew(clean, bias=False)) if n_valid >= 3 else None
    kurtosis = float(scipy_stats.kurtosis(clean, bias=False)) if n_valid >= 4 else None

    return {
        "n": int(n_valid),
        "missing": int(n_missing),
        "mean": round(mean, 6),
        "median": round(median, 6),
        "std": round(std, 6),
        "variance": round(variance, 6),
        "se_mean": round(se_mean, 6),
        "min": round(vmin, 6),
        "max": round(vmax, 6),
        "range": round(vmax - vmin, 6),
        "q1": round(q1, 6),
        "q3": round(q3, 6),
        "iqr": round(iqr, 6),
        "skewness": round(skewness, 6) if skewness is not None else None,
        "kurtosis": round(kurtosis, 6) if kurtosis is not None else None,
    }


# ---------------------------------------------------------------------------
# Resolve inputs from injected namespace
# ---------------------------------------------------------------------------

# Get data dictionary (injected by wrapper)
_data = data if "data" in dir() else {}  # noqa: F821

# Get variable names from the 'variables' slot (injected by wrapper)
try:
    _var_names = variables  # noqa: F821
except NameError:
    raise ValueError("Variable 'variables' is required - specify column names to analyze")

# Ensure it's a list
if isinstance(_var_names, str):
    _var_names = [_var_names]
elif not isinstance(_var_names, (list, tuple)):
    _var_names = list(_var_names)

if len(_var_names) == 0:
    raise ValueError("At least one variable must be specified")

# ---------------------------------------------------------------------------
# Process each variable
# ---------------------------------------------------------------------------

stats_out = {}
for var_name in _var_names:
    arr = _get_column(var_name, _data)
    stats_out[var_name] = _describe_column(arr)

result = {
    "variables": list(_var_names),
    "stats": stats_out,
}
